src/group_theory/amenable/mean.lean

```diff
@@ -0,0 +1,261 @@
+/-
+Copyright (c) 2022 Matthias Uschold. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matthias Uschold.
+-/
+
+import tactic
+import algebra.group.basic
+import data.real.basic
+import topology.continuous_function.bounded
+import topology.category.Top
+import topology.algebra.group
+import algebra.group.defs
+import algebra.order.hom.monoid
+
+/-!
+# Means on Groups
+
+In this file, we introduce means on groups,
+i.e. `averaging' maps that turn bounded (continuous) functions G → ℝ into a real number.
+
+
+## Main Definitions
+
+- `mean`              : Structure for means
+- `mean_pushforward`  : Pushing forward a mean on $G$ via a map $G → H$ yields a mean on $H$.
+
+## Implementation Notes
+
+We will ultimately need this notion to define amenability of groups via
+invariant means (in the file def_amenable).
+
+This file defines means by regarding all groups with their discrete topology,
+thus enabling us to use `bounded_continuous_function`.
+If you want to consider means on (non-discrete)
+topological groups, one needs to change some definitions.
+
+
+## References
+* <https://en.wikipedia.org/wiki/Amenable_group>
+* [C. Löh, *Geometric Group Theory*, Proposition 9.1.1][loeh17]
+
+
+## Tags
+mean
+-/
+
+open classical
+open bounded_continuous_function
+
+variables (G:Type*) [uniform_space G] [group G] [topological_group G]
+
+
+/-- A mean on a group-/
+structure mean := mk ::
+(lin_map : (bounded_continuous_function G ℝ) →ₗ[ℝ] ℝ)
+(normality : lin_map (bounded_continuous_function.const G (1:ℝ)) = 1)
+(monotonicity: monotone lin_map)
+
+
+instance : has_coe (mean G) ((bounded_continuous_function G ℝ) →ₗ[ℝ] ℝ) :=
+  {coe := mean.lin_map}
+
+
+/--Equality of means can be checked by evaluation -/
+@[ext]
+theorem ext {m n : mean G} (h: ∀ f, m f = n f) : m = n :=
+begin
+  cases m,
+  cases n,
+  simp,
+  ext,
+  exact h x,
+end
```

Comment:
I recommend instead implementing this as:
```lean
set_option old_structure_cmd true

structure mean extends
  bounded_continuous_function G ℝ →ₗ[ℝ] ℝ,
  one_hom (bounded_continuous_function G ℝ) ℝ,
  order_hom (bounded_continuous_function G ℝ) ℝ.

class mean_class (F : out_param Type*) (G : Type*) [uniform_space G] [group G] [topological_group G]
  extends
    semilinear_map_class F (ring_hom.id ℝ) (bounded_continuous_function G ℝ) ℝ,
    one_hom_class F (bounded_continuous_function G ℝ) ℝ,
    rel_hom_class F ((≤) : _ → bounded_continuous_function G ℝ → Prop) ((≤) : ℝ → ℝ → Prop).

instance mean.mean_class : mean_class (mean G) G :=
{ coe := mean.to_fun,
  coe_injective' := λ f g h, by { cases f, cases g, congr' },
  map_one := λ f, map_one f.to_one_hom,
  map_add := λ f, map_add f.to_linear_map,
  map_smulₛₗ := λ f, map_smul f.to_linear_map,
  map_rel := λ f, order_hom_class.monotone f.to_order_hom }
```
which will mean you don't need `mean_of_neg` below, the global `map_neg` should then work.

Have a look at the module docstring near `fun_like` for an explanation of the full pattern here; what I give above is just to get you started.

Reply:
A fantastic proposal, but I think we should use [lp](https://leanprover-community.github.io/mathlib_docs/analysis/normed_space/lp_space.html#lp) instead of `bounded_continuous_function` as I [said on Zulip](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Amenable.20Groups/near/300269636), and this file should not mention topology at all. We could later add amenability for topological groups but that involves L^infty for Haar measure which I imagine is much more complicated to work with, so it's better to work with `lp` and show the equivalence with the general definition when the topology is discrete.

And `mean_class` should probably be generalized to
```
class mean_class (F : out_param Type*) (V : Type*) [add_comm_group V] [module ℝ V] [has_one V]
  extends
    semilinear_map_class F ([ring_hom.id](http://ring_hom.id/) ℝ) V ℝ,
    one_hom_class F V ℝ,
    rel_hom_class F ((≤) : V → V → Prop) ((≤) : ℝ → ℝ → Prop).
```
to cover both the discrete and topological cases and potentially more.

Reply:
We should probably add `lp.has_one` instances when p=infty or [finite α].

Reply:
And in this particular case, we could merge `semilinear_map` and `order_hom` into `order_add_monoid_hom`; to show linearity, we can decompose a function in `lp` into a positive part and a negative part, and each part maps linearly by additivity, density of ℚ in ℝ, and monotonicity. This would make it easier to construct terms of type `mean`, and we can still show it fulfills `mean_class`. Not sure how much this can be generalized and in which file it should go ...

Reply:
Right, my comment was only "here's how to represent the structure you've currently chosen in our `fun_like` framework". It sounds like you're more qualified than me to comment on whether the structure that's currently here is the right one to begin with!

I'll note that I had a case recently where I wanted `semilinear_map_class` combined with `one_hom_class` (but not monotonicity); two examples that come to mind are `clifford_algebra.reverse` and `opposite.op`.

Reply:
> 



> I recommend instead implementing this as:
> 
> ```lean
> set_option old_structure_cmd true
> 
> structure mean extends
>   bounded_continuous_function G ℝ →ₗ[ℝ] ℝ,
>   one_hom (bounded_continuous_function G ℝ) ℝ,
>   order_hom (bounded_continuous_function G ℝ) ℝ.
> 
> class mean_class (F : out_param Type*) (G : Type*) [uniform_space G] [group G] [topological_group G]
>   extends
>     semilinear_map_class F (ring_hom.id ℝ) (bounded_continuous_function G ℝ) ℝ,
>     one_hom_class F (bounded_continuous_function G ℝ) ℝ,
>     rel_hom_class F ((≤) : _ → bounded_continuous_function G ℝ → Prop) ((≤) : ℝ → ℝ → Prop).
> 
> instance mean.mean_class : mean_class (mean G) G :=
> { coe := mean.to_fun,
>   coe_injective' := λ f g h, by { cases f, cases g, congr' },
>   map_one := λ f, map_one f.to_one_hom,
>   map_add := λ f, map_add f.to_linear_map,
>   map_smulₛₗ := λ f, map_smul f.to_linear_map,
>   map_rel := λ f, order_hom_class.monotone f.to_order_hom }
> ```
> 
> which will mean you don't need `mean_of_neg` below, the global `map_neg` should then work.

Thank you. I've done this now for the file `mean.lean`

> 
> Have a look at the module docstring near `fun_like` for an explanation of the full pattern here; what I give above is just to get you started.

I'm not so sure I could find this. Did you mean docs#fun_like or did you mean to point me to a different location?

Reply:
i.e. did you mean this file: [fun_like](https://leanprover-community.github.io/mathlib_docs/data/fun_like/basic.html#fun_like)

Reply:
> A fantastic proposal, but I think we should use [lp](https://leanprover-community.github.io/mathlib_docs/analysis/normed_space/lp_space.html#lp) instead of `bounded_continuous_function` as I [said on Zulip](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Amenable.20Groups/near/300269636), and this file should not mention topology at all. We could later add amenability for topological groups but that involves L^infty for Haar measure which I imagine is much more complicated to work with, so it's better to work with `lp` and show the equivalence with the general definition when the topology is discrete.


Ok, I'm ok with restricting to discrete groups for the moment. I'll try to implement it that way.